File: 2026-03-March-GFG-POTD/March 05 - Longest Substring with K Uniques/Solution.cpp
class Solution {
public:
    int longestKSubstr(string s, int k) {
        vector<int> freq(26, 0);
        int left = 0, right = 0;
        int distinct = 0;
        int maxLen = -1;

        while (right < s.length()) {
            char c = s[right];
            if (freq[c - 'a'] == 0)
                distinct++;

            freq[c - 'a']++;

            while (distinct > k) {
                char leftChar = s[left];
                freq[leftChar - 'a']--;

                if (freq[leftChar - 'a'] == 0)
                    distinct--;

                left++;
            }

            if (distinct == k)
                maxLen = max(maxLen, right - left + 1);

            right++;
        }

        return maxLen;
    }
};

File: 2026-03-March-GFG-POTD/March 05 - Longest Substring with K Uniques/Solution.py
class Solution:
    def longestKSubstr(self, s, k):
        freq = [0] * 26
        left = 0
        right = 0
        distinct = 0
        max_len = -1

        while right < len(s):
            idx = ord(s[right]) - ord('a')

            if freq[idx] == 0:
                distinct += 1

            freq[idx] += 1

            while distinct > k:
                left_idx = ord(s[left]) - ord('a')
                freq[left_idx] -= 1

                if freq[left_idx] == 0:
                    distinct -= 1

                left += 1

            if distinct == k:
                max_len = max(max_len, right - left + 1)

            right += 1

        return max_len
